### ISSUE TITLE
(Unsafe) type cast not taken into account for fields

### ISSUE THREAD

**Issue by yannl35133** [microsoft/pyright#2779]
**Describe the bug**
```python
class C:
    def foo(self):
        self.casted: int = None  # type: ignore

        - self.casted # Operator "-" not supported for type "None"
```

The type cast is not taken into account for the following type-checks.
Note that it does not happen if I replace `self.casted` by `casted`, this behaviour is apparently only present on fields

**VS Code extension or command-line**
VSCode extension v.1.1.202

**Comment by erictraut** [comment]
You haven't performed a cast in this case. You've declared that the type of the instance variable `self.casted` must contain only an `int`, then you've attempted to assign `None` to it, which is a type violation. Then you've suppressed the type violation error by including a `# type: ignore` comment. That doesn't change the fact that the variable `casted` is still declared to be type `int`. 

If your intent is for the `casted` variable to accept both `int` or `None`, then you should declared it as type `int | None`.

If you're new to type annotations in Python, I recommend reading [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md) which describes the basic concepts of type declarations, type inference, etc.

**Comment by yannl35133** [comment]
I guess cast was the wrong word, what I want is to have the type-checker believe that `self.casted` is of type `int`.
(There is a genuine use-case behind this, based on subtype with more methods)

**Comment by erictraut** [comment]
Is this what you have in mind?

```python
from typing import cast
class C:
    def foo(self):
        self.casted: int = cast(int, None)
        -self.casted
```

**Comment by yannl35133** [comment]
With class `A` and `B` subclass of `A`
```python
class C:
    def foo(self):
        self.b: B = B(sub_elements=[A()])  # type: ignore # sub_elements should be List[B]
        map_A_to_B(self.b.sub_elements) # self.b.sub_elements is now List[B], so self.b is B
        self.b.sub_elements[0].b_exclusive_method()
```

**Comment by erictraut** [comment]
When you assign a value to an expression, pyright applies type narrowing to the expression based on the value assigned. When you attempt to assign a value to an expression that violates the type rules, pyright emits an error (which you have suppressed with the ignore comment) and it then "narrows" the type of the LHS expression to the type of the RHS even though it violates the type declaration. For more information about type narrowing, please refer to the pyright documentation.

Your above example is incomplete, and I don't have enough context to tell what you're doing, so I can't offer much in the way of advice.

In any event, this isn't a bug in pyright. It's working as intended.

**Comment by yannl35133** [comment]
Why does it behave differently when it is a field and when it is a regular variable ?

**Comment by erictraut** [comment]
We don't apply false type narrowing for local variables when a type violation is detected. That's arguably an inconsistency, but it's done intentionally. Assignment-based type narrowing should really never be applied on type violations, but we changed this for field accesses to help pylance users who are not interested in type checking (and therefore don't care about type violations) but want the best possible completion suggestions. For users who are interested in static type checking, it shouldn't matter because their code will presumably not violate the type system. However, in your case, you're intentionally violating the type system and then suppressing the error, which is why you're hitting this case.

It sounds like you're interested in static type checking but are struggling to use it as it is intended — hence the need for casting and `# type: ignore` comments. You're fighting the type checker rather than having it work for you.

If you would like to provide a complete, self-contained sample, I can provide suggestions for how to annotate it such that you can avoid the use of these unsafe approaches.

**Comment by yannl35133** [comment]
Here is a quite faithful minimal representation of my problem
```python
from __future__ import annotations
from dataclasses import dataclass
from typing import List, Optional


class X:
    pass


class Y(X):
    @staticmethod
    def of_x(x: X) -> Y:
        raise NotImplementedError

    def foo(self):
        return None


@dataclass
class A:
    interesting_part: Optional[X]
    many_other_things = None
    sub_elements: List[A]

    def A_to_B_in_place(self: A):
        if self.interesting_part is not None:
            self.interesting_part = Y.of_x(self.interesting_part)
        for sub_elem in self.sub_elements:
            sub_elem.A_to_B_in_place()


# @dataclass
# class B(A):
#     interesting_part: Optional[Y]
#     sub_elements: List[B]


class C:
    def bar(self):
        self.a = A(None, [A(X(), [])])
        self.a.A_to_B_in_place()
        self.a.sub_elements[0].foo()

```

**Comment by erictraut** [comment]
A few thoughts here. 

First, you mentioned that you're doing in-place replacement rather than allocating new objects during the conversion, and you're doing this for performance reasons. I'm skeptical that this saves you much in terms of runtime performance. Have you actually profiled your code to determine whether this rather hacky approach is worth it? If you're really concerned about cutting cycles our of your code, you're using the wrong language. Python is not known to be a performant language.

Here's a potential solution that involves two separate classes. This is cleaner than an in-place conversion.

```python
_Node = TypeVar("_Node", bound="Tree[Any]")

@dataclass
class Tree(Generic[_Node]):
    sub_elements: List[_Node]

@dataclass
class A(Tree["A"]):
    interesting_part: Optional[X]
    many_other_things = None

    def convert_to_b(self) -> "B":
        interesting_part: Y | None = None
        if self.interesting_part is not None:
            interesting_part = Y.of_x(self.interesting_part)
        b_elements = [sub_elem.convert_to_b() for sub_elem in self.sub_elements]
        return B(b_elements, interesting_part)

@dataclass
class B(Tree["B"]):
    interesting_part: Optional[Y]

a = A([A([], X())], None)
b = a.convert_to_b()
assert b.sub_elements[0].interesting_part is not None
b.sub_elements[0].interesting_part.foo()
```

Second, if you're doing in-place replacement, then you shouldn't attempt to model it as two different classes. After all, it's not different classes that are instantiated at runtime, and modeling it as different classes will inevitably cause problems.

Here's a solution that involves a single class.

```python
@dataclass
class A:
    interesting_part: X | Y | None
    many_other_things = None
    sub_elements: List[A]

    def A_to_B_in_place(self: A):
        if self.interesting_part is not None:
            self.interesting_part = Y.of_x(self.interesting_part)
            self.interesting_part = None
        for sub_elem in self.sub_elements:
            sub_elem.A_to_B_in_place()


a = A(None, [A(X(), [])])
a.A_to_B_in_place()
assert isinstance(a.sub_elements[0].interesting_part, Y)
a.sub_elements[0].interesting_part.foo()
```

**Comment by yannl35133** [comment]
```python
from __future__ import annotations
from dataclasses import dataclass
from typing import Generic, List, Optional, TypeVar


class X:
    pass


class Y(X):
    @staticmethod
    def of_x(x: X) -> Y:
        raise NotImplementedError

    def foo(self):
        return None


T = TypeVar("T")


@dataclass
class A(Generic[T]):
    interesting_part: Optional[T] = None
    many = None
    other = None
    fields = None
    sub_elements: List[A[T]] = []

    def explore(self: A[Y]):
        if self.interesting_part is not None:
            self.interesting_part.foo()
        for sub_elem in self.sub_elements:
            sub_elem.explore()

    def AX_to_AY_in_place(self: A[X]):  # -> (self: A[Y], return: None)
        if self.interesting_part is not None:
            self.interesting_part = Y.of_x(self.interesting_part)
        for sub_elem in self.sub_elements:
            sub_elem.AX_to_AY_in_place()


class C:
    def bar(self):
        self.a: A[X] = A(sub_elements=[A(X(), [])])
        self.a.AX_to_AY_in_place()
        self.b = self.a  # type: ignore
        self.b: A[Y]
        self.b.explore()
```

This looks like an ok way of doing things, I just need a way to force a type and I can't find a way.
In your second solution, how would one check the type of all `interesting_part`s of the tree ?

**Comment by erictraut** [comment]
No, this won't work. You're once again modeling it as two different classes (`A[X]` and `A[Y]`) when your runtime code is using only one class. If you are insistent in doing in-place conversion (which I think is ill-advised and likely unnecessary), then you should model the types accordingly.

**Comment by yannl35133** [comment]
The reason I prefer in-place conversion is because I can't do things like `{a with interesting_field = Y.of_x(a.interesting_field)}` (bar the option part of the field), and I want to avoid the long code that is needed to list all fields.
Also, with this solution at least the interpreter won't notice the difference between the two types, as it shouldn't look for the type parameters.

Is there no way to force a type ?

Also, with your second solution, is there a way to make it accept `explore` ?

In any case, thanks for your time.
